Implemented method getSalaryInfo

DanaKhromenko opened the pull request:
I think I had missed the point.
There are some rows where the exceptions may appear, I marked them with "TODO"+comment.
I thought I need to add some exception class where I'll push some error-messages like "The data array is null" (can appear NullPointerException) or "The dateFrom value is invalid" (can appear DateTimeParseException).
But all of these exceptions are unchecked so I don't need to add a new checked-exception class I guess, and a programm should crash if incoming values are invalid.

DanaKhromenko commented:
Fixed Travis errors: "Line is longer than 100 symbols". Please read the previous comment!

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,105 @@
+        }
+
+        dateFrom = dateFrom.trim();
+        dateTo = dateTo.trim();
```
trim + валідація не потрібні так як в тебе тоді парсинг завалиться енівей якщо невалідні дані прийдуть, це трохи оверінженірінг

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,105 @@
+        LocalDate date1;
+        LocalDate date2;
+        try {
+            date1 = LocalDate.parse(dateFrom, DATE_FORMAT);
```
```suggestion
            from = LocalDate.parse(dateFrom, DATE_FORMAT);
```

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,105 @@
+            throw new RuntimeException("Can't parse DATE FROM, invalid value: " + dateFrom);
+        }
+        try {
+            date2 = LocalDate.parse(dateTo, DATE_FORMAT);
```
```suggestion
            to = LocalDate.parse(dateTo, DATE_FORMAT);
```

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,105 @@
+            throw new RuntimeException("Can't parse DATE TO, invalid value: " + dateTo);
+        }
+
+        for (int d = 0; d < data.length; d++) {
```
Буде значно простіше виглядати якщо ти зробиш зовн цикл для імен і для кожного імені будеш запускати внутр цикл по рядках data. Тоді заспліть рядок і подивись чи підходить час + дата. Якщо підходить сумуй зп зі всі рядків і додай в ріпорт

DanaKhromenko replied:
Но ведь в таком случае мы не 1 раз обходим циклом массив data[], а names[].length-раз. Разве это не повлияет на скорость исполнения кода? (не в данном конкретном случае, а в примерах в бОльшим количеством данных)

DanaKhromenko replied:
> Буде значно простіше виглядати якщо ти зробиш зовн цикл для імен і для кожного імені будеш запускати внутр цикл по рядках data. Тоді заспліть рядок і подивись чи підходить час + дата. Якщо підходить сумуй зп зі всі рядків і додай в ріпорт

Но ведь в таком случае мы не 1 раз обходим циклом массив data[], а names[].length-раз. Разве это не повлияет на скорость исполнения кода? (не в данном конкретном случае, а в примерах в бОльшим количеством данных)

DanaKhromenko replied:
> Буде значно простіше виглядати якщо ти зробиш зовн цикл для імен і для кожного імені будеш запускати внутр цикл по рядках data. Тоді заспліть рядок і подивись чи підходить час + дата. Якщо підходить сумуй зп зі всі рядків і додай в ріпорт

Я переделала, но всё же думаю, что здесь можно пожертвовать красотой и выиграть время исполнения кода.

RomanDubovskyi replied:
В тебе було аналогічно в рішенні: просто ти проходилась под даті і для кожного рядка ітерувала масив names. Це однакова складність алгоритму n^2

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,105 @@
 public class SalaryInfo {
+    private static final DateTimeFormatter DATE_FORMAT = DateTimeFormatter.ofPattern("dd.MM.yyyy");
+
     public String getSalaryInfo(String[] names, String[] data, String dateFrom, String dateTo) {
```
спочатку методи ти можеш одразу створити шапку для ріпорта по датай а потім коли будеш проходитись по names після сумування зп одної людини одразу додавати рядок в ріпорт

DanaKhromenko commented:
- Replaced dates trimming;
- Changed names of variables "date1" and "date2" with names "from" and "to";
- Changed the loops order so now I iterate names[] and inside I iterate data[]. 
- Appends report variable in the loop.

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,94 @@
+                    continue;
+                }
+
+                //checks if the date is reliable
```
код потрібно коментувати тільки в тих випадках якщо потрібно пояснити чому було прийнято неочевидне рішення. В більшості випадків потрібно старатись писати такий код, який буде зрозумілий і без коментування

RomanDubovskyi commented on src/main/java/core/basesyntax/SalaryInfo.java:
```diff
@@ -1,7 +1,94 @@
+                }
+
+                //parsing the salary data
+                int salaryIndex = currentData.lastIndexOf(' ');
```
для зручності я б просто розбив currentData через спліт по пробілу і звертався до даних по індексах результуючого масиву String

RomanDubovskyi reviewed (approved):
Стосовно швидкості: в тебе був аналогічний підхід спочатку: просто ти проходилась под даті і для кожного рядка ітерувала масив names. Це однакова алгоритмічна складність n^2. Тому те, що тут перероблено не дасть програшу в швидкості коду. 
![image](https://user-images.githubusercontent.com/58385506/124709873-aec5e480-df04-11eb-9c7c-bc155430204d.png)
